docs/Client.md

```diff
@@ -0,0 +1,989 @@
+
+Both arguments are optional; the method can be called in four different ways:
+
+```js
+client.destroy() // -> Promise
+client.destroy(new Error()) // -> Promise
+client.destroy(() => {}) // -> void
+client.destroy(new Error(), () => {}) // -> void
+```
+
+Arguments:
+
+* **error** `Error | null` (optional)
+* **callback** `() => void` (optional)
+
+Returns: `void | Promise<void>` - Only returns a `Promise` if no `callback` argument was passed
+
+#### Example - Request is aborted when Client is destroyed
+
+```js
+'use strict'
+const { createServer } = require('http')
+const { Client } = require('undici')
+
+const server = createServer((request, response) => {
+  response.end('undici')
+})
+
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  const request = client.request({
+    path: '/',
+    method: 'GET'
+  })
+
+  client.destroy()
+    .then(() => {
+      // Still waits for requests to complete
+      console.log('Client destroyed')
+      server.close()
+    })
+
+  // The request promise will reject with an Undici Client Destroyed error
+  request.catch(error => {
+    console.error(error)
+  })
+})
+```
+
+### `Client.dispatch(options, handlers)`
+
+This is the low level API which all the preceding APIs are implemented on top of.
+
+This API is expected to evolve through semver-major versions and is less stable than the preceding higher level APIs. It is primarily intended for library developers who implement higher level APIs on top of this.
+
+Arguments:
+
+* **options** `DispatchOptions`
+* **handlers** `DispatchHandlers`
+
+Returns: `void`
+
+#### Example 1 - Dispatch GET request
+
+```js
+'use strict'
+const { createServer } = require('http')
+const { Client } = require('undici')
+
+const server = createServer((request, response) => {
+  response.end('Hello, World!')
+})
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  const data = []
+
+  client.dispatch({
+    path: '/',
+    method: 'GET',
+    headers: {
+      'x-foo': 'bar'
+    }
+  }, {
+    onConnect: () => {
+      console.log('Connected!')
+    },
+    onError: (error) => {
+      console.error(error)
+    },
+    onHeaders: (statusCode, headers) => {
+      console.log(`onHeaders | statusCode: ${statusCode} | headers: ${headers}`)
+    },
+    onData: (chunk) => {
+      console.log('onData : chunk received')
+      data.push(chunk)
+    },
+    onComplete: (trailers) => {
+      console.log(`onComplete | trailers: ${trailers}`)
+      const res = Buffer.concat(data).toString('utf8')
+      console.log(`Data: ${res}`)
+      client.close()
+      server.close()
+    }
+  })
+})
+```
+
+#### Example 2 - Dispatch Upgrade Request
```

Comment:
@ronag @mcollina can y'all help me put together this example? I'm struggling to properly demonstrate an upgrade request using a http.Server

Reply:
I think you need a websocket server. Just base example on ws?

Reply:
I'm struggling to get the node.js http.server to fire the `'upgrade'` event.

I've tried a dispatch call with `method: 'CONNECT'` and `upgrade: 'websocket'` but it errors with `"SocketError: other side closed"`.

I also tried a GET request with connection and upgrade headers but those trigger an InvalidArgumentError 

🤷‍♂️

Reply:
OH I need to use `method: 'GET'` and `upgrade: 'websocket'` - got it now!